lib/Parse/Lexer.cpp

```diff
@@ -1477,6 +1471,22 @@ void Lexer::lexImpl() {
       } else {
         diagnose(CurPtr-1, diag::lex_invalid_character)
           .fixItReplaceChars(getSourceLoc(CurPtr-1), getSourceLoc(tmp), " ");
+
+        llvm::SmallString<64> buffer;
+        EncodeToUTF8(codepoint, buffer);
+        StringRef confusedChar = buffer.str();
+        uint32_t expectedCodepoint;
+        if ((expectedCodepoint =
+            confusable::tryConvertConfusableCharacterToASCII(codepoint))) {
+          std::string expected = std::string(1, (char)expectedCodepoint);
```

Comment:
Please use a SmallString here too (which implicitly converts to StringRef), so that we don't have any heap allocation.